nuxeo-coldstorage/src/main/java/org/nuxeo/coldstorage/service/ColdStorageServiceImpl.java

```diff
@@ -303,47 +320,65 @@ public DocumentModel requestRetrievalFromColdStorage(CoreSession session, Docume
                             documentModel),
                     SC_FORBIDDEN);
         }
-
-        try {
-            Blob coldContent = (Blob) documentModel.getPropertyValue(COLD_STORAGE_CONTENT_PROPERTY);
-            String key = getContentBlobKey(coldContent);
-            BlobUpdateContext updateContext = new BlobUpdateContext(key).withRestoreForDuration(restoreDuration);
-            Framework.getService(BlobManager.class).getBlobProvider(coldContent).updateBlob(updateContext);
-        } catch (IOException e) {
-            throw new NuxeoException(e);
+        BlobStatus blobStatus = getBlobStatus(documentModel);
+        boolean notify = false;
+        DocumentModel docResult = null;
+        if (isDownloadable(blobStatus)) {
+            documentModel.setPropertyValue(COLD_STORAGE_CONTENT_DOWNLOADABLE_UNTIL,
+                    Date.from(blobStatus.getDownloadableUntil()));
+            docResult = CoreInstance.doPrivileged(session, s -> {
+                // The retrieval is allowed for users with only READ access.
+                // It requires an unrestricted session to update ColdStorage metadata on the document
+                return s.saveDocument(documentModel);
+            });
```

Comment:
In order to avoid repeating this block, we could move it outside of the if/else if/else block, and turn `notify` into a Function<DocumentModel, Boolean> initialized to return false, and filled with `true` for the second block and the `checkIsRetrieved` code for the third block.